picolibc is incompatible with xcc / xcc-clang toolchains

**Describe the bug**

Compilation of picolibc module for picolibc tests fails with XCC on ADSP platforms. Followup to https://github.com/zephyrproject-rtos/zephyr/issues/54224.

A few of the flags used by GCC to build the picolibc module are not compatible with XCC. Different flags are observed to be not working for different versions of XCC toolchain.

`RI-2021.7-linux (xcc-clang)` + MTL:
```
clang-10: error: unknown argument: '-fno-printf-return-value'
clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]
```

`RG-2017.8-linux (xcc)` + CAVS15, 18, CAVS25:
```
xt-xcc ERROR parsing -frounding-math:  unknown flag
xt-xcc ERROR parsing -fsignaling-nans:  unknown flag
xt-xcc ERROR parsing -fno-stack-protector:  unknown flag
xt-xcc ERROR parsing -Warray-bounds:  unknown flag
xt-xcc ERROR parsing -ftls-model=local-exec:  unknown flag
xt-xcc ERROR parsing -fno-printf-return-value:  unknown flag
```

**To Reproduce**
```
export ZEPHYR_TOOLCHAIN_VARIANT=xcc
west build -p -b intel_adsp_cavs15 -T tests/kernel/common/kernel.common.picolibc
```

**Expected behavior**
Compilation

**Impact**
Annoyance

**Logs and console output**
See above.

**Environment**
 - OS: Ubuntu 20.04
 - Toolchain: XCC / XCC-clang (see above)
 - Recent main

ok, we need to keep this open because #54391 was not merged yet.

and I am getting this now with #54391 checked out:

```
clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]
[815/957] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libc/posix/regexec.c.obj
clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]
[816/957] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libc/posix/regcomp.c.obj
clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]
[940/957] Building C object zephyr/kernel/CMakeFiles/kernel.dir/main_weak.c.obj
FAILED: zephyr/kernel/CMakeFiles/kernel.dir/main_weak.c.obj
ccache /home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-clang -DKERNEL -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/nashif/zephyrproject/zephyr/kernel/include -I/home/nashif/zephyrproject/zephyr/arch/xtensa/include -I/home/nashif/zephyrproject/zephyr/include -I/home/nashif/zephyrproject/zephyr/build/zephyr/include/generated -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/nashif/zephyrproject/zephyr/drivers -isystem /home/nashif/zephyrproject/zephyr/build/modules/picolibc/picolibc/include -fno-strict-aliasing -Os -imacros /home/nashif/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -fno-common -g -gdwarf-4 -fcolor-diagnostics -fms-extensions -imacros /home/nashif/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-typedef-redefinition -Wno-pointer-sign -Wpointer-arith -Wno-sometimes-uninitialized -Wno-shift-overflow -Wno-missing-braces -Wno-self-assign -Wno-address-of-packed-member -Wno-unused-function -Wno-initializer-overrides -Wno-section -Wno-unknown-warning-option -Wno-unused-variable -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -Werror=implicit-int -fno-asynchronous-unwind-tables -Wno-vla -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr/tests/kernel/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/nashif/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -D_POSIX_C_SOURCE=200809 -std=c99 -MD -MT zephyr/kernel/CMakeFiles/kernel.dir/main_weak.c.obj -MF zephyr/kernel/CMakeFiles/kernel.dir/main_weak.c.obj.d -o zephyr/kernel/CMakeFiles/kernel.dir/main_weak.c.obj -c /home/nashif/zephyrproject/zephyr/kernel/main_weak.c
/home/nashif/zephyrproject/zephyr/kernel/main_weak.c:27:1: error: 'main' must return 'int'
void __weak main(void)
^~~~
int
1 error generated.
```


hmmm, did not see all of that last week. Will have to continue debugging tomorrow..

> clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]
> [815/957] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libc/posix/regexec.c.obj
> clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]
> [816/957] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libc/posix/regcomp.c.obj
> clang-10: warning: optimization flag '-fsignaling-nans' is not supported [-Wignored-optimization-argument]

This needs a picolibc fix to disable this flag as clang has no ability to correctly manage NaN values (meson picolibc already disables this flag, need to migrate that over to the cmake bits).

> /home/nashif/zephyrproject/zephyr/kernel/main_weak.c:27:1: error: 'main' must return 'int'
> void __weak main(void)
> ^~~~
> int

Oh, clang splits out warnings about `main` return type from other `main` related warnings. We need to add `-Wno-main-return-type`. #54525 has a proposed fix.

@nashif @keith-packard I cherrypicked both of those PRs and can replicate that the fix https://github.com/zephyrproject-rtos/zephyr/pull/54525 isn't working for us for some reason

```
$ west build -p -b intel_adsp_ace15_mtpm_sim -T tests/kernel/common/kernel.common.picolibc
...
[332/1047] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libc/string/memmove.c.obj
/home/laurenmu/zephyrproject/modules/lib/picolibc/newlib/libc/string/memmove.c:71:1: warning: malformed 'optimize' attribute '-fno-tree-loop-distribute-patterns' [-Wattribute-optimize]
__inhibit_loop_to_libcall
^
/home/laurenmu/zephyrproject/zephyr/build/modules/picolibc/picolibc/include/sys/cdefs.h:548:33: note: expanded from macro '__inhibit_loop_to_libcall'
  __attribute__ ((__optimize__ ("-fno-tree-loop-distribute-patterns")))
                                ^
1 warning generated.
[746/1047] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/math/sf_erf.c.obj
/tmp/sf_erf-35159d.s: Assembler messages:
/tmp/sf_erf-35159d.s:1185: Warning: value 0x4f80000000000000 truncated to 0x0
...
/home/laurenmu/zephyrproject/zephyr/subsys/testsuite/ztest/src/ztest_new.c:1029:1: error: 'main' must return 'int'
void main(void)
^~~~
int
```

even with `-Wno-main-return-type`

```
ccache /home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-clang -DKERNEL -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/nashif/zephyrproject/zephyr/kernel/include -I/home/nashif/zephyrproject/zephyr/arch/xtensa/include -I/home/nashif/zephyrproject/zephyr/include -I/home/nashif/zephyrproject/zephyr/build/zephyr/include/generated -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/nashif/zephyrproject/zephyr/drivers -isystem /home/nashif/zephyrproject/zephyr/build/modules/picolibc/picolibc/include -fno-strict-aliasing -Os -imacros /home/nashif/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -fno-common -g -gdwarf-4 -fcolor-diagnostics -fms-extensions -imacros /home/nashif/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-main-return-type -Wno-typedef-redefinition -Wno-pointer-sign -Wpointer-arith -Wno-sometimes-uninitialized -Wno-shift-overflow -Wno-missing-braces -Wno-self-assign -Wno-address-of-packed-member -Wno-unused-function -Wno-initializer-overrides -Wno-section -Wno-unknown-warning-option -Wno-unused-variable -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -Werror=implicit-int -fno-asynchronous-unwind-tables -Wno-vla -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr/tests/kernel/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/nashif/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -D_POSIX_C_SOURCE=200809 -std=c99 -MD -MT zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj -MF zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj.d -o zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj -c /home/nashif/zephyrproject/zephyr/kernel/init.c
/home/nashif/zephyrproject/zephyr/kernel/init.c:327:9: error: 'main' must return 'int'
        extern void main(void);
               ^~~~
               int
1 error generated.
```

> even with `-Wno-main-return-type`
> 
> ```
> ccache /home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-clang -DKERNEL -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/nashif/zephyrproject/zephyr/kernel/include -I/home/nashif/zephyrproject/zephyr/arch/xtensa/include -I/home/nashif/zephyrproject/zephyr/include -I/home/nashif/zephyrproject/zephyr/build/zephyr/include/generated -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/nashif/zephyrproject/zephyr/drivers -isystem /home/nashif/zephyrproject/zephyr/build/modules/picolibc/picolibc/include -fno-strict-aliasing -Os -imacros /home/nashif/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -fno-common -g -gdwarf-4 -fcolor-diagnostics -fms-extensions -imacros /home/nashif/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-main-return-type -Wno-typedef-redefinition -Wno-pointer-sign -Wpointer-arith -Wno-sometimes-uninitialized -Wno-shift-overflow -Wno-missing-braces -Wno-self-assign -Wno-address-of-packed-member -Wno-unused-function -Wno-initializer-overrides -Wno-section -Wno-unknown-warning-option -Wno-unused-variable -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -Werror=implicit-int -fno-asynchronous-unwind-tables -Wno-vla -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr/tests/kernel/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/nashif/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -D_POSIX_C_SOURCE=200809 -std=c99 -MD -MT zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj -MF zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj.d -o zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj -c /home/nashif/zephyrproject/zephyr/kernel/init.c
> /home/nashif/zephyrproject/zephyr/kernel/init.c:327:9: error: 'main' must return 'int'
>         extern void main(void);
>                ^~~~
>                int
> 1 error generated.
> ```

Ok, looks like there's something weird with the xtensa compiler. As I cannot use that compiler, I'm afraid someone with access will need to sort out what's wrong.

> > even with `-Wno-main-return-type`
> > ```
> > ccache /home/nashif/xtensa/XtDevTools/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-clang -DKERNEL -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/nashif/zephyrproject/zephyr/kernel/include -I/home/nashif/zephyrproject/zephyr/arch/xtensa/include -I/home/nashif/zephyrproject/zephyr/include -I/home/nashif/zephyrproject/zephyr/build/zephyr/include/generated -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/ace/include/intel_ace15_mtpm -I/home/nashif/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/include -I/home/nashif/zephyrproject/zephyr/subsys/testsuite/ztest/include -I/home/nashif/zephyrproject/zephyr/drivers -isystem /home/nashif/zephyrproject/zephyr/build/modules/picolibc/picolibc/include -fno-strict-aliasing -Os -imacros /home/nashif/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -fno-common -g -gdwarf-4 -fcolor-diagnostics -fms-extensions -imacros /home/nashif/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-main-return-type -Wno-typedef-redefinition -Wno-pointer-sign -Wpointer-arith -Wno-sometimes-uninitialized -Wno-shift-overflow -Wno-missing-braces -Wno-self-assign -Wno-address-of-packed-member -Wno-unused-function -Wno-initializer-overrides -Wno-section -Wno-unknown-warning-option -Wno-unused-variable -Wno-format-invalid-specifier -Wno-gnu -Wno-tautological-compare -Werror=implicit-int -fno-asynchronous-unwind-tables -Wno-vla -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr/tests/kernel/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/nashif/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/nashif/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -D_POSIX_C_SOURCE=200809 -std=c99 -MD -MT zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj -MF zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj.d -o zephyr/kernel/CMakeFiles/kernel.dir/init.c.obj -c /home/nashif/zephyrproject/zephyr/kernel/init.c
> > /home/nashif/zephyrproject/zephyr/kernel/init.c:327:9: error: 'main' must return 'int'
> >         extern void main(void);
> >                ^~~~
> >                int
> > 1 error generated.
> > ```
> 
> Ah, that's because we're not using GNU extensions to the standard.

--- a/cmake/compiler/clang/compiler_flags.cmake
+++ b/cmake/compiler/clang/compiler_flags.cmake
@@ -31,6 +31,7 @@ check_set_compiler_property(PROPERTY warning_base
                             -Wformat-security
                             -Wno-format-zero-length
                             -Wno-main
+                            -Wno-main-return-type
                             -Wno-unused-but-set-variable
                             -Wno-typedef-redefinition
                             -Wno-deprecated-non-prototype

